tests/schema.rs

```diff
@@ -624,6 +624,299 @@ fn test_equivalence_after_round_trip() {
+            "fields": [
+                {"name": "field_one", "type": "float"}
+            ]
+        }
+    }"#;
+    let schema_strs_first = [schema_str_1, schema_str_2];
+    let schema_strs_second = [schema_str_2, schema_str_1];
+    let schemas_first = Schema::parse_list(&schema_strs_first).expect("Test failed");
+    let schemas_second = Schema::parse_list(&schema_strs_second).expect("Test failed");
+
+    for schema_str in &[schema_str_1, schema_composite] {
+        let parsed = Schema::parse_str(&schema_str).expect("Test failed");
+        assert!(schemas_first.contains(&parsed));
+        assert!(schemas_second.contains(&parsed));
+        assert_eq!(schemas_first.len(), 2);
+        assert_eq!(schemas_second.len(), 2);
+    }
+}
+
+#[test]
+/// Test that schema composition resolves namespaces.
+fn test_parse_list_with_cross_deps_and_namespaces() {
+    let schema_str_1 = r#"{
+        "name": "A",
+        "type": "record",
+        "namespace": "namespace",
+        "fields": [
+            {"name": "field_one", "type": "float"}
+        ]
+    }"#;
+    let schema_str_2 = r#"{
+        "name": "B",
+        "type": "map",
+        "values": "namespace.A"
+    }"#;
+    let schema_composite = r#"{
+        "name": "B",
+        "type": "map",
+        "values": {
+            "name": "A",
+            "type": "record",
+            "namespace": "namespace",
+            "fields": [
+                {"name": "field_one", "type": "float"}
+            ]
+        }
+    }"#;
+    let schema_strs_first = [schema_str_1, schema_str_2];
+    let schema_strs_second = [schema_str_2, schema_str_1];
+    let schemas_first = Schema::parse_list(&schema_strs_first).expect("Test failed");
+    let schemas_second = Schema::parse_list(&schema_strs_second).expect("Test failed");
+
+    for schema_str in &[schema_str_1, schema_composite] {
+        let parsed = Schema::parse_str(&schema_str).expect("Test failed");
+        assert!(schemas_first.contains(&parsed));
+        assert!(schemas_second.contains(&parsed));
+        assert_eq!(schemas_first.len(), 2);
+        assert_eq!(schemas_second.len(), 2);
+    }
+}
+
+#[test]
+/// Test that schema composition fails on namespace errors.
+fn test_parse_list_with_cross_deps_and_namespaces_error() {
+    let schema_str_1 = r#"{
+        "name": "A",
+        "type": "record",
+        "namespace": "namespace",
+        "fields": [
+            {"name": "field_one", "type": "float"}
+        ]
+    }"#;
+    let schema_str_2 = r#"{
+        "name": "B",
+        "type": "map",
+        "values": "A"
+    }"#;
+
+    let schema_strs_first = [schema_str_1, schema_str_2];
+    let schema_strs_second = [schema_str_2, schema_str_1];
+    let _ = Schema::parse_list(&schema_strs_first).expect_err("Test failed");
+    let _ = Schema::parse_list(&schema_strs_second).expect_err("Test failed");
+}
+
+/// Return all permutations of an input slice
+fn permutations<T>(list: &[T]) -> Vec<Vec<&T>> {
```

Comment:
We may have been able to add [`itertools`](https://docs.rs/itertools/0.8.2/itertools/trait.Itertools.html#method.permutations) to our list of dev-dependencies (it's already a sub-dependency anyway), but this implementation is self-contained.
It's late here, but your permutations implementation looks good to me! 😄 Let's keep it.

Reply:
I felt guilty adding dependencies for a few tests so I thought a small amount of reinventing the wheel was worth it :sweat_smile:.